### DISCUSSION TITLE
TWS API running without errors, but is not placing any order #python #VS Code #Interactive Brokers #port7496 #ActiveXenabled

### DISCUSSION THREAD

**Discussion by ylvasvendsen** [community/community#132792] [Apps, API and Webhooks]
### Select Topic Area

Question

### Body

import pandas as pd
import numpy as np
from ibapi.client import EClient
from ibapi.wrapper import EWrapper
from ibapi.contract import Contract
from ibapi.order import Order
from ib_insync import IB, Stock, util
import time
from ibapi.ticktype import TickTypeEnum
import threading
import random
from sklearn.model_selection import TimeSeriesSplit
import logging
from datetime import datetime
from ibapi.client import EClient
from ib_insync import IB

# Connection details
port = 7496

print("\nConnecting to Interactive Brokers...")
ib = IB()
ib.connect("127.0.0.1", port, clientId=1)

# Logging configuration
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

tickers = ["LNG", "GOOGL", "AMZN", "AAPL", "CSCO", "INTC", "META", "MSFT", "NVDA", "TSLA"]

for ticker in tickers:
  stock = Stock(symbol=ticker, exchange="NASDAQ", currency="USD")
  # Rest of your code using the 'stock' variable

def buyStock(stock, quantity, limit_price):
    action = Order(
        action="BUY",
        totalQuantity=quantity,
        orderType="LMT",
        lmtPrice=limit_price,
        tif="GTC",
        outsideRth=True
    )
    order = ib.placeOrder(stock, action)

def sellStock(stock, quantity, limit_price):
    action = Order(
        action="SELL",
        totalQuantity=quantity,
        orderType="LMT",
        lmtPrice=limit_price,
        tif="GTC",
        outsideRth=True
    )
    order = ib.placeOrder(stock, action)

# Get real-time market data
data = ib.reqMktData(
    contract=stock,
    genericTickList="",
    snapshot=False,
    regulatorySnapshot=False
)

    def reqHistoricalData(self, contract, duration, barSize, keepUpToDate=False, chartOptions="", 
                          client=None, waitForConnection=True):
        """Requests historical data from TWS.

        See EClient.reqHistoricalData in the TWS API documentation for more details.

        Parameters
        contract (ib_insync.Contract): The contract to request data for.
        duration (str): The duration string (e.g. "1 M", "1 Y").
        barSize (str): The bar size string (e.g. "1 m", "1 d").
        keepUpToDate (bool): Whether to keep up to date with the latest data.
        chartOptions (str): Chart options string.
        client (ib_insync.IB, optional): The IB client instance. Defaults to None.
        waitForConnection (bool, optional): Whether to wait for connection before sending the request. Defaults to True.

        Returns
        ib_insync.BarContainer: A BarContainer object.
        """
        return super().reqHistoricalData(contract, duration, barSize, keepUpToDate, chartOptions, client, waitForConnection)

class IBapi(EWrapper, EClient):
    def __init__(self):
        EClient.__init__(self, self)
        self.contract = Contract()
        self.order = Order()
        self.orderId = None

def connect(self, host, port, clientId):
        try:
            EClient.connect(self, host, port, clientId)
        except Exception as e:
            print(f"Connection error: {e}")

    def nextValidId(self, orderId: int):
        self.orderId = orderId

    def nextId(self):
        self.orderId += 1
        return self.orderId

    def tickPrice(self, reqId, tickType, price, attrib):
        print(f"reqId: {reqId}, tickType: {TickTypeEnum.to_str(tickType)}, price: {price}, attrib: {attrib}")

    def tickSize(self, reqId, tickType, size):
        print(f"reqId: {reqId}, tickType: {TickTypeEnum.to_str(tickType)}, size: {size}")

    def error(self, reqId, errorCode, errorString):
        print("Error:", reqId, " ", errorCode, " ", errorString)

    def connect(self, host, port, clientId):
        try:
            EClient.connect(self, host, port, clientId)
        except Exception as e:
            print(f"Connection error: {e}")

    def placeOrder(self, orderId, contract, order):
        try:
            super().placeOrder(orderId, contract, order)
            print(f"Order placed: {orderId} - {contract.symbol} {order.action} {order.totalQuantity}")
        except Exception as e:
            print(f"Error placing order: {e}")

def calculate_aroon(data, period=25):
    aroon_up = 100 * data["High"].rolling(window=period + 1).apply(lambda x: x.argmax()) / period
    aroon_down = 100 * data["Low"].rolling(window=period + 1).apply(lambda x: x.argmin()) / period
    return aroon_up, aroon_down

def calculate_rsi(data, period=14):
    delta = data["Close"].diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
    rs = gain / loss
    rsi = 100 - (100 / (1 + rs))
    return rsi

def calculate_indicators(data, period=25):
    data["Aroon_Up"], data["Aroon_Down"] = calculate_aroon(data, period)
    data["RSI"] = calculate_rsi(data)
    data["MA50"] = data["Close"].rolling(window=50).mean()
    return data

def generate_trading_signals(data, stop_loss_percentage=0.05, take_profit_percentage=0.15):
    data["signal"] = 0
    data["stop_loss"] = data["Close"] * (1 - stop_loss_percentage)
    data['take_profit'] = data["Close"] * (1 + take_profit_percentage)
    buy_condition = (data["Aroon_Up"] > 70) & (data["Aroon_Down"] < 30) & (data["RSI"] <= 30) & (data["Close"] > data["MA50"])
    sell_condition = (data["RSI"] > 70) | (data["Close"] < data["stop_loss"]) | (data["Close"] > data["take_profit"])
    data.loc[buy_condition, "signal"] = 1
    data.loc[sell_condition, "signal"] = -1
    return data

def execute_trade(ib_client, signal, symbol, quantity, current_price, transaction_fee, trades):
    contract = Stock(symbol, "NASDAQ", "USD")
    
    order = Order()
    order.orderType = "LMT"
    order.totalQuantity = quantity
    order.action = "BUY" if signal == 1 else "SELL"
    order.lmtPrice = current_price * (1 + transaction_fee) if signal == 1 else current_price * (1 - transaction_fee)
    
    ib_client.placeOrder(ib_client.nextId(), contract, order)
    trades.append((symbol, order.action, quantity, current_price))
    
    logging.info(f"Trade executed: Symbol={symbol}, Action={order.action}, Quantity={quantity}, Price={current_price:.2f}")

def get_historical_data(ib, symbol, duration, bar_size):
    contract = Stock(symbol, "NASDAQ", "USD")
    bars = ib.reqHistoricalData(
        contract,
        endDateTime='',
        durationStr=duration,
        barSizeSetting=bar_size,
        whatToShow="MIDPOINT",
        useRTH=True,
        formatDate=1
    )
    df = util.df(bars)
    df["Date"] = pd.to_datetime(df["date"])
    return df

def filter_low_std_stocks(symbols, ib, duration="1 Y", bar_size="1 day", threshold=0.15):
    low_std_stocks = []
    for symbol in symbols:
        data = get_historical_data(ib, symbol, duration, bar_size)
        std_dev = data["Close"].std()
        if std_dev < threshold:
            low_std_stocks.append(symbol)
    return low_std_stocks

def optimize_parameters(data, stop_loss_range, take_profit_range):
    best_params = (0.05, 0.15)
    best_performance = -np.inf
    tscv = TimeSeriesSplit(n_splits=5)
    
    for stop_loss in stop_loss_range:
        for take_profit in take_profit_range:
            performance = 0
            for train_index, test_index in tscv.split(data):
                train, test = data.iloc[train_index], data.iloc[test_index]
                signals = generate_trading_signals(train, stop_loss, take_profit)
                performance += backtest_performance(signals, test)
            
            avg_performance = performance / tscv.get_n_splits()
            if avg_performance > best_performance:
                best_performance = avg_performance
                best_params = (stop_loss, take_profit)
    
    return best_params

def backtest_performance(signals, test_data):
    # Implement a simple backtest to measure performance
    return np.random.rand()  # Replace with actual backtest logic

def trading_bot(symbols, quantity, stop_loss_percentage, take_profit_percentage, max_iterations, transaction_fee):
    ib = IBapi()
    ib.connect("127.0.0.1", port, 1)
    threading.Thread(target=ib.run).start()
    time.sleep(1)  # Give time for the connection to be established
    
    trades = []
    symbols = filter_low_std_stocks(symbols, ib)
    
    for symbol in symbols:
        daily_data = get_historical_data(ib, symbol, "1 Y", "1 day")
        intraday_data = get_historical_data(ib, symbol, "1 W", "15 mins")
        
        daily_data = calculate_indicators(daily_data)
        intraday_data = calculate_indicators(intraday_data)
        
        stop_loss_range = np.arange(0.01, 0.11, 0.01)
        take_profit_range = np.arange(0.05, 0.20, 0.01)
        best_params = optimize_parameters(daily_data, stop_loss_range, take_profit_range)
        stop_loss_percentage, take_profit_percentage = best_params
        
        daily_signals = generate_trading_signals(daily_data, stop_loss_percentage, take_profit_percentage)
        intraday_signals = generate_trading_signals(intraday_data, stop_loss_percentage, take_profit_percentage)
        
        combined_signals = pd.concat([daily_signals, intraday_signals]).drop_duplicates("Date")
        
        last_trade_time = None
        trade_interval = pd.Timedelta(minutes=15)
        
        for i in range(len(combined_signals)):
            signal = combined_signals["signal"].iloc[i]
            current_price = combined_signals["Close"].iloc[i]
            current_time = combined_signals["Date"].iloc[i]
            
            if signal != 0 and (last_trade_time is None or current_time - last_trade_time >= trade_interval):
                execute_trade(ib, signal, symbol, quantity, current_price, transaction_fee, trades)
                last_trade_time = current_time
    
    ib.disconnect()
    return trades

# Example usage
symbols = ["LNG", "GOOGL", "AMZN", "AAPL", "CSCO", "INTC", "META", "MSFT", "NVDA", "TSLA"]
quantity = 1
stop_loss_percentage = 0.10
take_profit_percentage = 0.02
max_iterations = 500
transaction_fee = 0.001

trades = trading_bot(symbols, quantity, stop_loss_percentage, take_profit_percentage, max_iterations, transaction_fee)
print(trades)

**Comment by github-actions** [comment]
🕒 **Discussion Activity Reminder** 🕒

This Discussion has been labeled as dormant by an automated system for having no activity in the last 60 days. Please consider one the following actions:

1️⃣ Close as Out of Date: If the topic is no longer relevant, close the Discussion as `out of date` at the bottom of the page.

2️⃣ Provide More Information: Share additional details or context — or let the community know if you've found a solution on your own.

3️⃣ Mark a Reply as Answer: If your question has been answered by a reply, mark the most helpful reply as the solution.

Note: This dormant notification will only apply to Discussions with the `Question` label. To learn more, see our [recent announcement](https://github.com/orgs/community/discussions/70478).

Thank you for helping bring this Discussion to a resolution! 💬
